zproject/backends.py

```diff
@@ -846,6 +848,22 @@ def sync_full_name_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUs
                 raise ZulipLDAPError(e.msg)
             do_change_full_name(user_profile, full_name, None)
 
+    def sync_mail_from_ldap(
+        self, current_mail: str, ldap_user: _LDAPUser, realm: Realm
+    ) -> Optional[UserProfile]:
+        if (
+            settings.AUTH_LDAP_FOREIGN_KEY in ldap_user.attrs
+            and UserProfile.objects.filter(
+                ext_auth_id__ldap=ldap_user.attrs[settings.AUTH_LDAP_FOREIGN_KEY][0], realm=realm
+            ).exists()
```

Comment:
Probably we should fetch the user once with `.get`, rather than doing the `exists` check and then fetching it in the next line. And we should compare the emails before calling a `do_` function.